### PR TITLE
Storage operations refactor.

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

The operations `sstore` and `sload` were being used incorrectly. We were passing in byte pointers, when we should have been using words.

### How was it fixed?

Despite the fact that these operations take word pointers, I think we should continue using byte pointers to express storage values. This allows us to store composite/array types more gas efficiently, since we can reference multiple values inside of a single word. We also get to keep the same internal encoding of values in storage and memory. 

To deal with this, the storage pointers expressed in bytes are simply converted to word and byte offset pairs when it comes time to load or manipulate storage. One tricky part of this is that values cannot span multiple words in storage. In memory, you can load/store a word at any byte address, meaning you can have a packed array of 20 byte values, without having to work with multiple words. This is not the case in storage, we can only address words. So for example, the 2nd address in an address array would occupy two words. Trying to deal with values spanning multiple words would be too impractical, so we should store values in such a way that no single value spans multiple words. For now, this is only an issue with address arrays, but it is something that we will also need to consider when storing structs more efficiently.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Update storage operations and add thorough testing
- [x] Update everything that uses these storage operations
- [x] Fix encoding of address arrays
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)

- [x] Clean up commit history

### CODE REVIEW

File: common/src/utils/keccak.rs [hunk 1]
```diff
@@ -7,6 +7,24 @@ pub fn get_full_signature(content: &[u8]) -> String {
     get_partial_signature(content, 32)
 }
 
+pub fn get_partial_signature_padded(content: &[u8], size: usize) -> String {
```
Comment (g-r-a-n-t) [inline common/src/utils/keccak.rs:10]: rename these functions

File: compiler/src/yul/mappers/expressions.rs [hunk 2]
```diff
@@ -495,90 +495,90 @@ mod tests {
             .to_string()
     }
 
-    #[test]
```
Comment (g-r-a-n-t) [inline compiler/src/yul/mappers/expressions.rs:498]: readd tests

File: compiler/src/yul/operations/data.rs [hunk 3]
```diff
@@ -12,33 +12,39 @@ use yultsur::*;
 /// The returned expression evaluates to a 256 bit value.
 pub fn sload<T: FeSized>(typ: T, sptr: yul::Expression) -> yul::Expression {
     let size = literal_expression! { (typ.size()) };
-    expression! { sloadn([sptr], [size]) }
+    let word_ptr = expression! { div([sptr.clone()], 32) };
+    let bytes_offset = expression! { mod([sptr], 32) };
+    expression! { sloadn([word_ptr], [bytes_offset], [size]) }
```
Comment (g-r-a-n-t) [inline compiler/src/yul/operations/data.rs:17]: there should be a separate function `bytes_sloadn` or something

File: compiler/src/yul/operations/data.rs [hunk 4]
```diff
@@ -12,33 +12,39 @@ use yultsur::*;
...
 }
 
 /// Stores a 256 bit value in storage.
 pub fn sstore<T: FeSized>(typ: T, sptr: yul::Expression, value: yul::Expression) -> yul::Statement {
     let size = literal_expression! { (typ.size()) };
-    statement! { sstoren([sptr], [value], [size]) }
+    let word_ptr = expression! { div([sptr.clone()], 32) };
+    let bytes_offset = expression! { mod([sptr], 32) };
+    statement! { sstoren([word_ptr], [bytes_offset], [size], [value]) }
```
Comment (g-r-a-n-t) [inline compiler/src/yul/operations/data.rs:25]: bytes_sstoren

File: compiler/src/yul/runtime/functions/data.rs [hunk 5]
```diff
@@ -66,20 +66,23 @@ pub fn load_data_string() -> yul::Statement {
 pub fn mcopys() -> yul::Statement {
     function_definition! {
         function mcopys(mptr, sptr, size) {
-            (let offset := 0)
-            (for { } (lt((add(offset, 32)), size)) { }
+            (let mptr_offset := 0)
+            (let sptr_offset := 0)
+            (for { } (lt((add(mptr_offset, 32)), size)) { }
```
Comment (g-r-a-n-t) [inline compiler/src/yul/runtime/functions/data.rs:71]: see if this test expression can be cleaned up at all for each of the copying functions

File: compiler/src/yul/runtime/functions/data.rs [hunk 6]
```diff
@@ -198,11 +204,20 @@ pub fn mstoren() -> yul::Statement {
 /// Stores a value in storage, only modifying the given size (0 < size <= 32).
 pub fn sstoren() -> yul::Statement {
     function_definition! {
-        function sstoren(ptr, val, size) {
-            (let size_bits := mul(8, size))
-            (let left := shl((sub(256, size_bits)), val))
-            (let right := shr(size_bits, (sload((add(ptr, size))))))
-            (sstore(ptr, (or(left, right))))
+        function sstoren(word_ptr, bytes_offset, bytes_size, val) {
```
Comment (g-r-a-n-t) [inline compiler/src/yul/runtime/functions/data.rs:207]: try to clean this function up

File: compiler/src/yul/runtime/functions/data.rs [hunk 7]
```diff
@@ -240,3 +256,15 @@ pub fn ternary() -> yul::Statement {
         }
     }
 }
+
+pub fn set_zero() -> yul::Statement {
+    function_definition! {
+        function set_zero(start_bit, end_bit, val) -> result {
+            (let left_shift := sub(256, start_bit))
```
Comment (g-r-a-n-t) [inline compiler/src/yul/runtime/functions/data.rs:263]: left_shift_dist and right_shift_dist would be better names

File: compiler/src/yul/runtime/functions/mod.rs [hunk 8]
```diff
@@ -27,6 +27,7 @@ pub fn std() -> Vec<yul::Statement> {
         data::dualkeccak256(),
```
Comment (g-r-a-n-t) [inline compiler/src/yul/runtime/functions/mod.rs:27]: this function should just be renamed

File: compiler/src/yul/mappers/expressions.rs [hunk 9]
```diff
@@ -20,7 +20,7 @@ use fe_analyzer::{
     Context,
```
Comment (g-r-a-n-t) [inline compiler/src/yul/mappers/expressions.rs:20]: make sure address arrays in storage work

File: compiler/src/yul/mappers/expressions.rs [hunk 10]
```diff
@@ -459,15 +459,18 @@ fn expr_attribute_self(
 }
 
 /// Converts a storage nonce into a pointer based on the keccak256 hash
+///
+/// Pointers created here have the last byte set to zero. This is to ensure that
+/// our byte pointer sits at the start of a word (32 | `ptr` ).
 pub fn nonce_to_ptr(nonce: usize) -> yul::Expression {
-    let ptr = get_full_signature(nonce.to_string().as_bytes());
+    // set the last byte to `0x00` to ensure our pointer sits at the start of a word
```
Comment (cburgdorf) [inline compiler/src/yul/mappers/expressions.rs:466]: I'm having trouble to understand this. Let's say the full keccak of my `nonce`  would be `0xfe615a44e52a090747a78488bb7f44500a00175bbb8ff698965c2bf8ff06aeba`. As I understand it, we simply set the last byte to `0` so instead our pointer becomes `0xfe615a44e52a090747a78488bb7f44500a00175bbb8ff698965c2bf8ff060000`. I don't see how this changes anything.

>to ensure our pointer sits at the start of a word

I don't make the connection to this statement. Maybe I have to read on with the rest of the PR to understand it better...

**Review by cburgdorf — COMMENTED** [review]
I think it would be good to discuss this PR in our meeting today. I think I understand the general rational and also this part.

>In memory, you can load/store a word at any byte address, meaning you can have a packed array of 20 byte values, without having to work with multiple words. This is not the case in storage, we can only address words. So for example, the 2nd address in an address array would occupy two words.

However, this PR doesn't seem to only affect the size in storage but the size in memory, too.

If I have the following Fe code:

```
    pub def bar(someone: address):
        folks: address[1]
        folks[0] = someone
```

Then on current master it will compile to this:

```
function $$bar($someone) {
    let $folks := alloc(20)
    mstoren(add($folks, mul(0, 20)), $someone, 20) 
}
```

But with this PR it will compile to this:

```
function $$bar($someone) {
    let $folks := alloc(32)
    mstoren(add($folks, mul(0, 32)), 32, $someone) 
}
```

So, it seems that we would be wasting space in memory even if we don't need to.

**Review by cburgdorf — APPROVED** [review]

**Pull request merged**